### PR TITLE
Rework the public API

### PR DESCRIPTION
**Pull request by dbatyai**
Related to #4186.

Some notable changes:
  - The term 'Error' now strictly refers to native Error objects defined in
    the ECMA standard, which are ordinary objects. All other uses of
    'error' or 'error reference' where the term refers to a thrown value is
    now called 'exception'.

  - Simplified the naming scheme of many String API functions. These functions
    will now also take an 'encoding' argument to specify the desired
    encoding in which to operate.

  - Removed the substring-copy-to-buffer functions. These functions
    behaved awkwardly, as they use character index to specify the
    start/end positions, and were mostly used incorrectly with byte
    offsets instead. The functionality can still be replicated with
    other functions if necessary.

  - String-to-buffer functions will no longer fail if the buffer is not
    sufficiently large, the string will instead be cropped.

  - Fixed the usage of the '_sz' prefix in many API functions. The term
    'sz' means zero-terminated string in hungarian notation, this was
    used incorrectly in many cases.

  - Renamed most of the public API functions to have shorter, more on-point
    names, rather than the often too long descriptive names. Functions are now
    also grouped by the type of value they operate on, where this makes
    sense.

### CODE REVIEW

File: docs/02.API-REFERENCE.md [hunk 1]
```diff
@@ -7731,22 +7468,22 @@ The behaviour of the Proxy can be specialized with an options argument.
...
     when it is no longer needed.
 
 **Prototype**
 
 ```c
 jerry_value_t
-jerry_create_special_proxy (const jerry_value_t target,
-                            const jerry_value_t handler,
-                            uint32_t options);
+jerry_proxy_special (const jerry_value_t target,
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:7478]: `jerry_proxy_custom` and `jerry_proxy_object_options_t`-> `jerry_proxy_custom_behavior_t`

File: docs/02.API-REFERENCE.md [hunk 2]
```diff
@@ -8203,30 +7852,30 @@ jerry_create_regexp (const jerry_char_t *pattern_p, uint16_t flags);
...
   ...
 
-  jerry_release_value (regexp);
+  jerry_value_free (regexp);
 }
 ```
 
 
-## jerry_create_regexp_sz
+## jerry_regexp_sz_sz
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:7864]: jerry_regexp

File: docs/02.API-REFERENCE.md [hunk 3]
```diff
@@ -8156,38 +7805,38 @@ main (void)
...
 
-## jerry_create_regexp
+## jerry_regexp_sz
 
 **Summary**
 
-Returns a `jerry_value_t` RegExp object or an error, if the construction of the object fails.
+Returns a RegExp object or an exception, if the construction of the object fails.
 Optional flags can be set using [jerry_regexp_flags_t](#jerry_regexp_flags_t).
 These flags can be combined together with the binary OR operator or used on their own as enum values.
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:7831]: add asci note.

File: docs/02.API-REFERENCE.md (hunk 2 above)
Comment (rerobika) [reply docs/02.API-REFERENCE.md:7864]: how about construct it from string?

File: docs/02.API-REFERENCE.md [hunk 4]
```diff
@@ -8374,27 +8023,27 @@ For the new object the type of the TypedArray (see: [jerry_typedarray_type_t](#j
...
 
 **Prototype**
 
 ```c
 jerry_value_t
-jerry_create_typedarray_for_arraybuffer_sz (jerry_typedarray_type_t type_name,
-                                            const jerry_value_t arraybuffer,
-                                            jerry_length_t byte_offset,
-                                            jerry_length_t length);
+jerry_typedarray_with_buffer_range (jerry_typedarray_type_t type_name,
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:8043]: type_name -> type

Comment (rerobika) [reply docs/02.API-REFERENCE.md:8043]: range -> span.

File: docs/02.API-REFERENCE.md [hunk 5]
```diff
@@ -9021,18 +8670,18 @@ The receiver is passed as the `this` argument for getters, and the receiver
...
 
 **Prototype**
 
 ```c
 jerry_value_t
-jerry_get_own_property (const jerry_value_t obj_val,
-                        const jerry_value_t prop_name_val,
-                        const jerry_value_t receiver_val,
-                        bool *found_p);
+jerry_object_get_own (const jerry_value_t obj_val,
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:8681]: jerry_object_find_own

File: docs/02.API-REFERENCE.md [hunk 6]
```diff
@@ -1082,18 +1081,18 @@ typedef struct
...
-- [jerry_native_pointer_release_references](#jerry_native_pointer_release_references)
-- [jerry_native_pointer_set_reference](#jerry_native_pointer_set_reference)
+- [jerry_object_set_native_ptr](#jerry_object_set_native_ptr)
+- [jerry_object_get_native_ptr](#jerry_object_get_native_ptr)
+- [jerry_object_delete_native_ptr](#jerry_object_delete_native_ptr)
+- [jerry_native_ptr_init](#jerry_native_ptr_init)
+- [jerry_native_ptr_free](#jerry_native_ptr_free)
+- [jerry_native_ptr_set](#jerry_native_ptr_set)
 
 ## jerry_object_property_foreach_t
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:1091]: jerry_object_property_foreach_cb_t

File: docs/02.API-REFERENCE.md [hunk 7]
```diff
@@ -10494,14 +10136,14 @@ Iterate over all objects available in the engine.
...
-*Note*: Values obtained in `foreach_p` must be retained using [jerry_acquire_value](#jerry_acquire_value).
+*Note*: Values obtained in `foreach_p` must be retained using [jerry_value_copy](#jerry_value_copy).
 
 **Prototype**
 
 ```c
 bool
-jerry_objects_foreach (jerry_objects_foreach_t foreach_p,
-                       void *user_data_p);
+jerry_foreach_live_object (jerry_foreach_live_object_t foreach_p,
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:10145]: cb suffix

File: docs/02.API-REFERENCE.md [hunk 8]
```diff
@@ -10621,9 +10263,9 @@ main (void)
 
 **See also**
 
-- [jerry_objects_foreach_t](#jerry_objects_foreach_t)
+- [jerry_foreach_live_object_t](#jerry_foreach_live_object_t)
 
-## jerry_objects_foreach_by_native_info
+## jerry_foreach_live_object_with_info
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:10268]: cb

File: docs/02.API-REFERENCE.md [hunk 9]
```diff
@@ -10969,7 +10545,7 @@ Create an external JerryScript engine context.
 
 ```c
 jerry_context_t *
-jerry_create_context (uint32_t heap_size,
+jerry_context_create (uint32_t heap_size,
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:10548]: jerry_context

File: docs/02.API-REFERENCE.md [hunk 10]
```diff
@@ -11525,28 +11100,28 @@ main (void)
...
 **See also**
 
-- [jerry_get_backtrace_from](#jerry_get_backtrace_from)
-- [jerry_create_external_function](#jerry_create_external_function)
+- [jerry_backtrace_from](#jerry_backtrace_from)
+- [jerry_function_external](#jerry_function_external)
 
 
-## jerry_backtrace_capture
+## jerry_backtrace
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:11107]: jerry_backtrace_capture

File: docs/02.API-REFERENCE.md [hunk 11]
```diff
@@ -13232,7 +12804,7 @@ jerry_get_typedarray_buffer (jerry_value_t value,
 Parses a JSON string creating a JavaScript value. The behaviour is equivalent with
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md:12804]: add note about encoding

File: docs/02.API-REFERENCE.md [hunk 12]
```diff
@@ -3711,45 +3666,45 @@ jerry_create_abort_from_value (jerry_value_t value, bool release);
...
 **See also**
 
 - [jerry_value_t](#jerry_value_t)
-- [jerry_get_value_from_error](#jerry_get_value_from_error)
-- [jerry_create_error_from_value](#jerry_create_error_from_value)
+- [jerry_exception_value](#jerry_exception_value)
+- [jerry_exception](#jerry_exception)
 
-## jerry_create_error_from_value
+## jerry_exception
```
Comment (rerobika) [inline docs/02.API-REFERENCE.md]: Outdated.

**Review by rerobika — APPROVED** [review]
LGTM

File: targets/curie_bsp/jerry_app/quark/main.c [hunk 13]
```diff
@@ -13,59 +13,60 @@
...
-#include "infra/reboot.h"
 #include "infra/log.h"
-#include "infra/time.h"
+#include "infra/reboot.h"
 #include "infra/system_events.h"
 #include "infra/tcmd/handler.h"
-
-#include "cfw/cfw.h"
+#include "infra/time.h"
 /* Watchdog helper */
```
Comment (galpeter) [inline targets/curie_bsp/jerry_app/quark/main.c:28]: This is now an incorrect comment imho

**Review by galpeter — APPROVED** [review]
lgtm

File: docs/02.API-REFERENCE.md [hunk 14]
```diff
@@ -328,7 +324,7 @@ Flags for [jerry_exec_snapshot](#jerry_exec_snapshot) functions:
  - JERRY_SNAPSHOT_EXEC_COPY_DATA - copy snapshot data into memory (see below)
  - JERRY_SNAPSHOT_EXEC_ALLOW_STATIC - allow executing static snapshots
  - JERRY_SNAPSHOT_EXEC_LOAD_AS_FUNCTION - load snapshot as function instead of executing it
- - JERRY_SNAPSHOT_EXEC_HAS_RESOURCE - `resource_name` field is valid
+ - JERRY_SNAPSHOT_EXEC_HAS_RESOURCE - `source_name` field is valid
```
Comment (lygstate) [inline docs/02.API-REFERENCE.md:327]: Mayebe JERRY_SNAPSHOT_EXEC_HAS_SOURCE?

**Comment by lygstate** [comment]
Suggest the in new public API
rename
```
void jerry_port_print_char (char c);
```
for more detailed message printing
```
void jerry_port_print_string (const char *s, size_t len); 
```
For example, for UTF8 string, jerry_port_print_char will not able to print a Unicode character
in a single function call.

**Comment by dbatyai** [comment]
I wouldn't necessarily replace `jerry_port_print_char`, it has it's uses in it's current form. However, I agree that another port function could be added that can be used to print a string. I'm currently working on updating the port and ext API-s, so I can add this in the next PR.

As far as I can tell the `debug_buffer` array isn't strictly necessary at the moment either, it's only there to buffer the data over the debugger connection. In my opinion it could be removed regardless.

**Comment by lygstate** [comment]
jerry_port_print_string can be implement with jerry_port_print_char,
I've already have the  work that implement jerry_port_print_string across exist platforms.
Maybe I can do the work first.

**Comment by lygstate** [comment]
what's the difference betwen jerry_length_t and jerry_size_t? can we merge them into jerry_size_t?
I do not see any difference between these, and there is mixed usage of them
for example

jerry-core\include\jerryscript-core.h
```c
jerry_value_t jerry_typedarray_buffer (const jerry_value_t value, jerry_size_t *byte_offset, jerry_size_t *byte_length);
```
jerry-core\api\jerryscript.c
```c
/**
 * Get the underlying ArrayBuffer from a TypedArray.
 *
 * Additionally the byteLength and byteOffset properties are also returned
 * which were specified when the TypedArray was created.
 *
 * Note:
 *     the returned value must be freed with a jerry_value_free call
 *
 * @return ArrayBuffer of a TypedArray
 *         TypeError if the object is not a TypedArray.
 */
jerry_value_t
jerry_typedarray_buffer (const jerry_value_t value, /**< TypedArray to get the arraybuffer from */
                         jerry_length_t *byte_offset, /**< [out] byteOffset property */
                         jerry_length_t *byte_length) /**< [out] byteLength property */
{
  jerry_assert_api_enabled ();

#if JERRY_BUILTIN_TYPEDARRAY
  if (!ecma_is_typedarray (value))
  {
    return jerry_throw_sz (JERRY_ERROR_TYPE, ecma_get_error_msg (ECMA_ERR_OBJECT_IS_NOT_A_TYPEDARRAY));
  }

  ecma_object_t *array_p = ecma_get_object_from_value (value);
  uint8_t shift = ecma_typedarray_get_element_size_shift (array_p);

  if (byte_length != NULL)
  {
    *byte_length = (jerry_length_t) (ecma_typedarray_get_length (array_p) << shift);
  }

  if (byte_offset != NULL)
  {
    *byte_offset = (jerry_length_t) ecma_typedarray_get_offset (array_p);
  }

  ecma_object_t *arraybuffer_p = ecma_typedarray_get_arraybuffer (array_p);
  ecma_ref_object (arraybuffer_p);
  return jerry_return (ecma_make_object_value (arraybuffer_p));
#else /* !JERRY_BUILTIN_TYPEDARRAY */
  JERRY_UNUSED (value);
  JERRY_UNUSED (byte_length);
  JERRY_UNUSED (byte_offset);
  return jerry_throw_sz (JERRY_ERROR_TYPE, ecma_get_error_msg (ECMA_ERR_TYPED_ARRAY_NOT_SUPPORTED));
#endif /* JERRY_BUILTIN_TYPEDARRAY */
} /* jerry_typedarray_buffer */

```

**Pull request merged**